### ISSUE TITLE
cannot play certain 128 kbps MP3 streams

### ISSUE THREAD

**Issue by edmundlaugasson** [owncloud/music#966]
These streams are:
- http://icecast.pereraadio.ee/Tartu
- http://icecast.pereraadio.ee/Tallinn
- http://icecast.pereraadio.ee/Kuressaare

E.g. [mpv](https://mpv.io/) plays at CLI in Linux these streams.

At the same time Music plays the MP3-stream https://icecast.err.ee/klassikaraadiokorge.mp3 but does not play same stream with .m3u extension https://icecast.err.ee/klassikaraadiokorge.mp3.m3u (this URL actually exist) - same applies to Opus streams. You can check these streams at https://icecast.err.ee/

**Comment by paulijar** [comment]
Hello. I managed to play the first three streams on the Music app. The thing about these is probably that you access your cloud via a https URL but those streams use plain http. By default, all modern browsers block such mixed access. You need to separately allow "less secure content on this page" from somewhere within your browser settings.

On the other hand, the first thing to note about the m3u streams is that the servers providing those need to be separately white-listed in the `config/config.php` file of your cloud. For example, this kind of configuration would do the trick on the example case:
```
  'music.allowed_radio_hls_src' =>
  array (
    0 => 'https://icecast.err.ee'
  ),
```
However, this doesn't seem to be enough on that particular URL. This is because this server doesn't define the response headers allowing the cross-origin access. Hence, any modern browser enforcing the CORS checks cannot play it embedded to any other page except for https://icecast.err.ee/. I don't know any easy way around this.

**Comment by paulijar** [comment]
Upon closer examination, I found out that https://icecast.err.ee/klassikaraadiokorge.mp3.m3u is not a HLS-stream after all. Instead, it is just a playlist file which points to the actual stream URL https://icecast.err.ee/klassikaraadiokorge.mp3. A more common approach is to use PLS-type playlist file for this, but Music currently doesn't support that, either.

However, the support has now been implemented and the next version of Music will be able to play streams defined using PLS or M3U playlists. Parsing the playlists happens on the backend, so playlist URL not defining the CORS headers won't be problem either, as long as the actual audio stream defines the correct headers. 

Also, the handling of HLS streams has been heavily modified and those will no longer need the white-listing described above.

**Comment by edmundlaugasson** [comment]
Tried recent Music 1.5.2 and it could not play mentioned Pereraadio links, although at CLI the mpv could play. Still also Opus streams do not play.
Those playlists are not under my power to change. I can only change link, what to add. If editing provided link will work. But in case of Pereraadio I could not use anything else than provided links. At CLI with mpv these play but not in Music 1.5.2.

**Comment by edmundlaugasson** [comment]
I could get Vikerraadio to play, when removed the m3u part from URL but similar trick did not work with other URL's. Tried to add certain extensions in order to get them to play, did not succeed. Certainly I tested before at CLI, whether that URL plays at all and then tried in Music app.

**Comment by paulijar** [comment]
So are you accessing your cloud via HTTPS? And have you enabled the "insecure content" on you browser? On Chrome, the setting looks like this:
![image](https://user-images.githubusercontent.com/8565946/182696457-4ebc0589-f21f-4706-9dc8-99d6a9a10a6f.png)

**Comment by edmundlaugasson** [comment]
Using Chromium in Linux and there I have to specify the URL to allow. Added http://icecast.pereraadio.ee/, tried also icecast.pereraadio.ee and [*.]pereraadio.ee but none worked via Music app. At the same time if I open in same browser e.g. http://icecast.pereraadio.ee/Tartu then it plays fine.

**Comment by paulijar** [comment]
I just tested with the latest Chromium 106.0.5229.0 on Win 10. Also there, http://icecast.pereraadio.ee/Tartu was playing fine, after I just enabled the Insecure content on Chromium and reloaded the page.

When the playback fails for you, what can you see in the browser developer console (opens with the F12 key)?

**Comment by edmundlaugasson** [comment]
That was good idea to look at developer console. As I mentioned, I don't see an option to just allow insecure content. I must insert server address on which I allow insecure content. I've added:
- https://icecast.pereraadio.ee
- http://icecast.pereraadio.ee
- https://stun.nextcloud.com as it came out from logs
- https://my.serveraddress.ee
- http://my.serveraddress.ee

For all added addresses I even allowed redirections and pop-ups + sound. Insecure content was already allowed.

When I tested with address http://icecast.pereraadio.ee/Tartu.m3u then I got such error log:
[error-log-with-Tartu-m3u.log](https://github.com/owncloud/music/files/9294874/error-log-with-Tartu-m3u.log)


When tested with http://icecast.pereraadio.ee/Tartu  then I got such error log:
[error-log-with-Tartu.log](https://github.com/owncloud/music/files/9294872/error-log-with-Tartu.log)

As the last address http://icecast.pereraadio.ee/Tartu is actually correct, then we should use it. I just tried to mitigate possible issue, that playlist file extension might be needed.

Anyway, in last log is referred the URL in message:
`The page at 'https://my.serveraddress.ee/apps/music/#/radio' was loaded over HTTPS, but requested an insecure element 'http://icecast.pereraadio.ee/Tartu'. This request was automatically upgraded to HTTPS, For more information see https://blog.chromium.org/2019/10/no-more-mixed-messages-about-https.html`

So it seems, that such allowing insecure content does just not work as I don't know, how to prevent such http->https autoupgrade. I'm still wondering, how you still succeeded to play. Is it really the newest development version of Chromium (you had v106.x, I have v104.x) or is it difference between MS Windows and GNU/Linux version of Chromium.

I found [the Chromium release schedule](https://chromiumdash.appspot.com/schedule). Also on same site there are the [branches table of Chromium](https://chromiumdash.appspot.com/branches) and I have the 5112 of version 104, which is the latest currently.

Also we can check, which versions on which platforms are available:
- [MS Windows](https://chromiumdash.appspot.com/releases?platform=Windows)
- [GNU/Linux](https://chromiumdash.appspot.com/releases?platform=Linux)
- [macOS](https://chromiumdash.appspot.com/releases?platform=Mac)
- [Android](https://chromiumdash.appspot.com/releases?platform=Android)
- [iOS](https://chromiumdash.appspot.com/releases?platform=iOS)

So we truly see, that on August 9, 2022 there is 106.0.5229.0 for MS Windows available. For GNU/Linux at the same time is 106.0.5216.6 from August 5, 2022 available.

So, I used the address https://download-chromium.appspot.com/ to download the very latest version from trunk - 106.0.5230.0 was available at that moment for GNU/Linux, [build revision 1033232](https://crrev.com/1033232) - when I downloaded, there was message, that 4 minutes back was updated.... So, very-very latest version...
Then tested with that version 106.0.5230.0 and still got exactly same error messages. Although I'm using Chromium in Estonian language, the setting is the same. I guess also you can to allow insecure content after you entered appropriate URL as allowing is connected with URL.

**Comment by edmundlaugasson** [comment]
Then I finally changed addresses, when allowing insecure content:
- https://[*.]pereraadio.ee
- http://[*.]pereraadio.ee
- https://[*.]serveraddress.ee
- http://[*.]serveraddress.ee

... where _serveraddress.ee_ is the address, where my cloud storage instance is installed and used.

I could not wonder, when I noticed same mixed content warning:
`Mixed Content: The page at 'https://my.server.ee/apps/music/#/radio' was loaded over HTTPS, but requested an insecure audio file 'http://icecast.pereraadio.ee:8000/Tartu'. This content should also be served over HTTPS.`
... where http was requested over http and I noticed the end of sentence: _....this content should be also served over HTTPS_ **but still it finally started to play!**
I used Chromium v104.0.5112.79

**Comment by paulijar** [comment]
Okay, glad you got it working finally. But I'm kind of puzzled about how you needed to list some addresses for insecure content. Where did you list these?

For me, all I did was click this button:
![image](https://user-images.githubusercontent.com/8565946/183921115-403756ed-7bfc-4ac2-a4c5-de1967ef1ce3.png)

Then selected "Site Settings":
![image](https://user-images.githubusercontent.com/8565946/183921465-25472d31-7a83-4fd0-ba66-3f41472e1db6.png)

And switched the "Insecure content" to "Allow":
![image](https://user-images.githubusercontent.com/8565946/183922431-f11b1063-44a4-49f9-b02f-7146d8e1928c.png)

Before allowing the Insecure content, I got the same warning `The page at 'https://my.serveraddress.ee/apps/music/#/radio' was loaded over HTTPS, but requested an insecure element 'http://icecast.pereraadio.ee/Tartu'. This request was automatically upgraded to HTTPS, For more information see https://blog.chromium.org/2019/10/no-more-mixed-messages-about-https.html` and the error when the HTTPS-upgraded URL didn't work.

And after allowing it, I can see the same warning `Mixed Content: The page at 'https://my.server.ee/apps/music/#/radio' was loaded over HTTPS, but requested an insecure audio file 'http://icecast.pereraadio.ee:8000/Tartu'. This content should also be served over HTTPS.` but it still plays.

> Certainly, we don't know, how long that party can last until Chromium developers decide to completely turn off that option as they warned on [certain article in Chromium blog](https://blog.chromium.org/2019/10/no-more-mixed-messages-about-https.html) already on October 3, 2019.

I quickly glimpsed that article and I got the understanding that they were just going to disable the mixed content *by default*. And that they have done, that's the reason why we need to specifically allow the "Insecure content". But I didn't see any statement that they would plan to completely remove this option.

**Comment by edmundlaugasson** [comment]
Basically it is the same place - I noticed the address _chrome://settings/content/insecureContent_ on address bar - this link should work regardless in which language the Chromium to use. There I added both my and streaming server addresses and both http, https addresses with wildcard replacing subdomains.
I also hope, that Chromium developers will not turn off http.

**Comment by paulijar** [comment]
Okay. When I enable the Insecure content from the Site settings, then a row appears automatically in the view chrome://settings/content/insecureContent. That row looks like https://my.cloud.domain.fi:443 while I access my cloud with a URL like https://my.cloud.domain.fi/nextcloud. And I have no need to add anything else, like the URLs of the streams.

**Comment by paulijar** [comment]
As you can now listen to the stations, I think we can close this issue. Also, the newly released Music v1.6.0 can now play also the URL https://icecast.err.ee/klassikaraadiokorge.mp3.m3u. Furthermore, any HLS-type radio stations no longer need the separate white-listing I mentioned in my [first reply](https://github.com/owncloud/music/issues/966#issuecomment-1105791916).
